src/Radicle/Internal/Doc.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE TemplateHaskell       #-}
+
+module Radicle.Internal.Doc where
+
+import           Protolude hiding (Any)
+
+import           Codec.Serialise (Serialise(..))
+import           Data.Copointed (Copointed(..))
+import qualified Data.Default as Default
+import           Data.Pointed (Pointed(..))
+import           Language.Haskell.TH.Quote
+import           Text.Pandoc
+
+type Blocks = [Block]
+type Inlines = [Inline]
+
+data Pan = Blocks Blocks | Inlines Inlines
+  deriving (Eq, Ord, Show, Read)
+
+instance Semigroup Pan where
+  Blocks x <> Blocks y = Blocks (x <> y)
+  Inlines x <> Inlines y = Inlines (x <> [Space] <> y)
+  Blocks x <> Inlines y = Blocks (x <> [Para y])
+  Inlines x <> Blocks y = Blocks (Para x : y)
+
+class ToPan a where
+  pan :: a -> Pan
+
+instance ToPan Pan where
+  pan = identity
+
+blocks :: ToPan a => a -> Blocks
+blocks x = case pan x of
+  Blocks bs  -> bs
+  Inlines is -> [Para is]
+
+instance ToPan Blocks where
+  pan = Blocks
+
+instance ToPan a => ToPan (Maybe a) where
+  pan Nothing  = Inlines []
+  pan (Just x) = pan x
+
+data Described a = Desc (Maybe Pan) a
+  deriving (Eq, Ord, Generic, Show, Read)
+
+instance Serialise (Described a) where
+  encode = notImplemented
+  decode = notImplemented
+
+instance ToPan a => ToPan (Described a) where
+  pan (Desc (Just d) x) = d <> pan x
+  pan (Desc _ x)        = pan x
+
+data Value
+  = Fun Function
+  | String
+  | Number
+  | Boolean
+  | Atom
+  | Doc
+  | Any
+  | Other
+  deriving (Eq, Ord, Generic, Show, Read)
+
+instance ToPan Value where
+  pan = \case
+      Fun f -> pan f
+      String -> typ "string"
+      Number -> typ "number"
+      Boolean -> typ "boolean"
+      Atom -> typ "atom"
+      Doc -> typ "doc"
+      Any -> typ "any"
+      Other -> Inlines []
+    where
+
+typ :: Text -> Pan
```

Comment:
I think you misformated this `where`

Reply:
Lol it's actually also used in `instance ToPan Function`.